Reorientation

AKuederle opened the pull request.

DMegaritis committed:
First implementation of helper function to rotate acceleration according to quaternions

DMegaritis committed:
First implementation of helper function to rotate acceleration according to quaternions

DMegaritis committed:
First implementation of filter_signals_100Hz function high and low band butterworth

DMegaritis committed:
First implementation of filter_signals_100Hz function high and low band butterworth

DMegaritis committed:
utils implemented with gaitmap functions using rotations instead of quaternions

DMegaritis committed:
utils implemented with gaitmap functions using rotations instead of quaternions

DMegaritis committed:
Improved implementation of utils

DMegaritis committed:
Improved implementation of utils

DMegaritis committed:
ValueError added

DMegaritis committed:
ValueError added

DMegaritis committed:
First implementation of CorrectSensorOrientationDynamic

DMegaritis committed:
First implementation of CorrectSensorOrientationDynamic

DMegaritis committed:
Simplified and more robust since data are already sliced before this function is called

DMegaritis committed:
Simplified and more robust since data are already sliced before this function is called

DMegaritis committed:
first implementation of CorrectOrientationSensorAxes.py

DMegaritis committed:
first implementation of CorrectOrientationSensorAxes.py

DMegaritis committed:
Updated dockstrings

DMegaritis committed:
Updated dockstrings

DMegaritis committed:
resolved warnings

DMegaritis committed:
resolved warnings

DMegaritis committed:
updated filters using the mobgap toolbox

DMegaritis committed:
updated filters using the mobgap toolbox

DMegaritis committed:
Improved selection of non-GS signal

DMegaritis committed:
Improved selection of non-GS signal

DMegaritis committed:
Updated data types

DMegaritis committed:
Updated data types

DMegaritis committed:
Updated quat multiplication in line with MATLAB

DMegaritis committed:
Updated quat multiplication in line with MATLAB

DMegaritis committed:
Updated method of butter.

DMegaritis committed:
Updated method of butter.

DMegaritis committed:
Updated method of butter

DMegaritis committed:
Updated method of butter

DMegaritis committed:
Updated mobgap filters

DMegaritis committed:
Updated mobgap filters

DMegaritis committed:
Fixed bugs

DMegaritis committed:
Fixed bugs

DMegaritis committed:
Merge remote-tracking branch 'origin/main' into Reorientation

DMegaritis committed:
Added Reorientation after updating to mobgap

DMegaritis committed:
Added Reorientation after updating to mobgap

DMegaritis committed:
Added init

DMegaritis committed:
Added init

DMegaritis committed:
Updated renaming

DMegaritis committed:
Updated renaming

DMegaritis committed:
Madgwick testing on lab data

DMegaritis committed:
Madgwick testing on lab data

DMegaritis committed:
Madgwick testing on lab data

DMegaritis committed:
Madgwick testing on lab data

DMegaritis committed:
Madgwick application on whole array

DMegaritis committed:
Madgwick application on whole array

DMegaritis committed:
Applying madwick in the whole array

DMegaritis committed:
Applying madwick in the whole array

DMegaritis committed:
Update constant (th) to reflect data in m/s2

DMegaritis committed:
Update constant (th) to reflect data in m/s2

DMegaritis committed:
Signal correction for non GS areas moved out of the main loop to allow for signal correction when bouts are =< 2

DMegaritis committed:
Signal correction for non GS areas moved out of the main loop to allow for signal correction when bouts are =< 2

DMegaritis committed:
Signal correction for non GS areas moved out of the main loop to allow for signal correction when bouts are =< 2

DMegaritis committed:
Signal correction for non GS areas moved out of the main loop to allow for signal correction when bouts are =< 2

DMegaritis committed:
Updated initial orientation according to Matlab function (UpdateIMUslt2)

DMegaritis committed:
Updated initial orientation according to Matlab function (UpdateIMUslt2)

DMegaritis committed:
Fixed bugs

DMegaritis committed:
Fixed bugs

DMegaritis committed:
Updated dockstrings

DMegaritis committed:
Updated dockstrings

DMegaritis committed:
Updated dockstrings and tidied code

DMegaritis committed:
Updated dockstrings and tidied code

AKuederle reviewed (commented):
Some general comments @DMegaritis:

Can we add the minimal test-cases to the repo? As part of the tests/example_data folder. Have a look on how example data is handled for the other algos.

And for the algorithms, can you start the refactoring process into the class structure? I also added some direct comments to some parts of the algorithms

AKuederle commented on mobgap/Reorientation/CorrectOrientationSensorAxes.py:
```diff
@@ -0,0 +1,117 @@
+        GS = pd.DataFrame(columns=['Start', 'Stop'])
+
+        if n > 2:
+            for i in gs.index:
```
YOu could use the the `iter_gs` util to make that a easier to read

AKuederle commented on mobgap/Reorientation/CorrectOrientationSensorAxes.py:
```diff
@@ -0,0 +1,117 @@
+
+        ind_noGS = groupfind(gsLabel == 0)
+
+        for i in range(len(ind_noGS[:, 0])):
```
As above iter_gs might help

AKuederle commented on mobgap/Reorientation/CorrectSensorOrientationDynamic.py:
```diff
@@ -0,0 +1,175 @@
+    - Parameters are expressed in the units used in mobgap,specifically, we use m/s^2 instead of g.
+    """
+
+    acc = data.iloc[:, 0:3]
```
Named access

AKuederle commented on mobgap/Reorientation/CorrectSensorOrientationDynamic.py:
```diff
@@ -0,0 +1,175 @@
+    # implementation (UpdateIMUslt2). This choice was made since the first "initial orientation"
+    # specified in the Matlab implementation is overwritten in the UpdateIMUslt2 function.
+    
+    mad = MadgwickAHRS(beta=0.1, initial_orientation=np.array([-0.0863212587840360, -0.639984676042049, 
```
Do we know where this initial orientation comes from? Does this initial orientation make sense?

DMegaritis replied:
This is the initial orientation specified in the respective MATLAB function. 
To my understanding, the initial orientation describes a rotation where the human bends forwards slightly less that 90 degrees.

AKuederle commented on mobgap/Reorientation/CorrectSensorOrientationDynamic.py:
```diff
@@ -0,0 +1,175 @@
+    # applying the madgwick algo to the entire data array
+    data = pd.concat((chosenacc, chosengyr), axis=1)
+    mad = mad.estimate(data, sampling_rate_hz=sampling_rate_hz)
+    quaternion = mad.orientation_.iloc[1:, :]
```
Positional access

DMegaritis replied:
Yes, this removes the first instance which is the initial orientation making the "quaternion" identical in length to the data

AKuederle commented on mobgap/Reorientation/CorrectOrientationSensorAxes.py:
```diff
@@ -0,0 +1,117 @@
+
+                avm = np.mean(av_filt1[l1:l2])
+
+                if avm >= th:
```
Also in this case nothing happens right? So why are we overwriting the corIMUData

DMegaritis replied:
Yes, this case suggests that the accelerometer signal looks "normal"

AKuederle commented on mobgap/Reorientation/CorrectSensorOrientationDynamic.py:
```diff
@@ -0,0 +1,175 @@
+    quaternion = quaternion.to_numpy()
+
+    # Adjust quaternion from x, y, z, w to w, x, y, z
+    quaternion = quaternion[:, [3, 0, 1, 2]]
```
I would highly recommed to use the rotation object directly instead of custom quternion math. The scipy implemntation is more accurate and propbly much much faster

DMegaritis replied:
I have tried using functions from scipy.spatial.transform and gaitmap.utils.rotations but the outcomes are different to the matlab. 
So I have implemented custom functions to replicate the source code of the respective matlab functions. We can use existing libraries for efficiency but the outcomes would be different.

AKuederle commented on mobgap/Reorientation/CorrectSensorOrientationDynamic.py:
```diff
@@ -0,0 +1,175 @@
+
+    # Principal component analysis
+    pca_acc = PCA()
+    pca_acc.fit(chosenacc)
```
Just use `fit_transform` instead of manually transforing into the new coordinate system

AKuederle commented on mobgap/Reorientation/CorrectSensorOrientationDynamic.py:
```diff
@@ -0,0 +1,175 @@
+    newgyr = chosengyr.values @ pca_coef_gyr.T
+
+    if np.mean(newacc[:, 0]) < 0:
+        newacc[:, 0] = -newacc[:, 0]
```
Named access

DMegaritis replied:
This is a NumPy array. I compared the output between this and the matlab and it is exactly identical so I am sure I use the same data. Do you think it is better to convert to a pd dataframe and use named access?

AKuederle commented on mobgap/Reorientation/CorrectSensorOrientationDynamic.py:
```diff
@@ -0,0 +1,175 @@
+    sig4 = (av[:, 1] - np.mean(av[:, 1])) / np.std(av[:, 1])
+
+    # Assigning av_pca and gyr_pca
+    av_pca = av_magpca.copy()
```
Why are we creatingnew variables?

DMegaritis replied:
This was not needed, I removed it.

AKuederle commented on mobgap/Reorientation/CorrectSensorOrientationDynamic.py:
```diff
@@ -0,0 +1,175 @@
+    av_pca = av_magpca.copy()
+    gyr_pca = gyr_magpca.copy()
+
+    # Calculating correlations
```
Can you add some comments on what is happening here? I am confused

DMegaritis replied:
I added comments in the whole code below this point

AKuederle commented on mobgap/Reorientation/CorrectSensorOrientationDynamic.py:
```diff
@@ -0,0 +1,175 @@
+    gyr_magpca[:, 0] = newgyr[:, 0]        # Yaw
+
+    # Standardization
+    sig1 = (av[:, 2] - np.mean(av[:, 2])) / np.std(av[:, 2])
```
Instead of perfoming all of these transformations per axis, you can probaly do this vectorized on the entire frames.

AKuederle commented on mobgap/Reorientation/CorrectSensorOrientationDynamic.py:
```diff
@@ -0,0 +1,175 @@
+
+    if abs(cor1) > abs(cor2):
+        if cor1 > 0:
+            av_pca[:, 2] = newacc[:, 1]  # AP
```
Named access!

DMegaritis replied:
Updated to dataframes so named access is used now

AKuederle commented on mobgap/Reorientation/CorrectSensorOrientationDynamic.py:
```diff
@@ -0,0 +1,175 @@
+
+    filter_chain = [("savgol", savgol)]
+
+    av_pca_filt[:, 2] = chain_transformers(av_pca[:, 2], filter_chain, sampling_rate_hz=sampling_rate_hz)
```
Same as above. Can probably all be vectorized

AKuederle commented on mobgap/Reorientation/CorrectSensorOrientationDynamic.py:
```diff
@@ -0,0 +1,175 @@
+        gyr_pca_final[:, 1] = gyr_pca[:, 2]
+
+    IMU_corrected = np.concatenate((av_pca_final, gyr_pca_final), axis=1)
+    IMU_corrected = pd.DataFrame(IMU_corrected)
```
This dataframe does not have the original names and missing the index. I would suggest to try to keep the dataframe for most of the operation, instead of switching to numpy dataframes

DMegaritis replied:
I updated the function to use dataframes so named access is used now everywhere

AKuederle reviewed (commented):
Some general comments @DMegaritis:

Can we add the minimal test-cases to the repo? As part of the tests/example_data folder. Have a look on how example data is handled for the other algos.

And for the algorithms, can you start the refactoring process into the class structure? I also added some direct comments to some parts of the algorithms

DMegaritis committed:
Named instead of positional access

DMegaritis committed:
Named instead of positional access

DMegaritis committed:
Named instead of positional access

DMegaritis committed:
Named instead of positional access

DMegaritis committed:
Added else clause to handle short data properly

DMegaritis committed:
Added else clause to handle short data properly

DMegaritis committed:
corIMUdata is not a copy of data.

DMegaritis committed:
corIMUdata is not a copy of data.

DMegaritis committed:
Use of iter_gs

DMegaritis committed:
Use of iter_gs

DMegaritis committed:
fit_transform replaced manual transformation

DMegaritis committed:
fit_transform replaced manual transformation

DMegaritis committed:
Vectorised standardisation

DMegaritis committed:
Vectorised standardisation

DMegaritis committed:
Vectorised filtering and standardisation

DMegaritis committed:
Vectorised filtering and standardisation

DMegaritis committed:
Added comments next to operations.
Removed transpose from dot product since only one column is present.
Removed creation of new variable (copy of av_magpca and gyr_magpca) since they were not needed

DMegaritis committed:
Added comments next to operations.
Removed transpose from dot product since only one column is present.
Removed creation of new variable (copy of av_magpca and gyr_magpca) since they were not needed

DMegaritis committed:
Updated comments

DMegaritis committed:
Updated comments

DMegaritis committed:
acceleration function returns a pd.DataFrame

DMegaritis committed:
acceleration function returns a pd.DataFrame

DMegaritis committed:
Updated arrays to pd.dataframes so named access can be used

DMegaritis committed:
Updated arrays to pd.dataframes so named access can be used

DMegaritis committed:
Corrections on named access

DMegaritis committed:
Corrections on named access

DMegaritis committed:
Updated comments of operations

DMegaritis committed:
Updated comments of operations

DMegaritis committed:
Updated comments of operations

DMegaritis committed:
Updated comments of operations

DMegaritis committed:
Corrected 'n' referring to the number of walking bouts

DMegaritis committed:
Corrected 'n' referring to the number of walking bouts

DMegaritis committed:
Added ValueError and cleaned

DMegaritis committed:
Added ValueError and cleaned

DMegaritis committed:
Added ValueError and cleaned

DMegaritis committed:
Added ValueError and cleaned

DMegaritis committed:
Updated init

DMegaritis committed:
Updated init

DMegaritis committed:
Updated import statements

DMegaritis committed:
Updated import statements

DMegaritis committed:
acceleration: IMU shape should be 6 cols and not 3

DMegaritis committed:
acceleration: IMU shape should be 6 cols and not 3

DMegaritis committed:
Updated value errors

DMegaritis committed:
Updated value errors

DMegaritis committed:
Updated value errors

DMegaritis committed:
Updated value errors

DMegaritis committed:
Rebased to class structure

DMegaritis committed:
Rebased to class structure

DMegaritis committed:
Removed old structure

DMegaritis committed:
Removed old structure

DMegaritis committed:
Name update

DMegaritis committed:
Name update

DMegaritis committed:
Removed testing of mad

DMegaritis committed:
Removed testing of mad

DMegaritis committed:
Updated init to new structure

DMegaritis committed:
Updated init to new structure

DMegaritis committed:
Updated dockstrings

DMegaritis committed:
Updated dockstrings

DMegaritis commented:
> Some general comments @DMegaritis:
> 
> Can we add the minimal test-cases to the repo? As part of the tests/example_data folder. Have a look on how example data is handled for the other algos.
> 
> And for the algorithms, can you start the refactoring process into the class structure? I also added some direct comments to some parts of the algorithms



I made several amendments following your comments @AKuederle, and I improved other parts as well. I refactored the main function to class structure. Do you think it is good to refactor the secondary function as well (correct_sensor_orientation_dynamic)? 
Regarding testing, we could add the data under 'example_data' and some tests under 'tests' when I have the files ready. I need to do more work on this because the work I have done in creating 8 'wrong' files does not always lead to walking bouts.

DMegaritis reviewed (commented):
I made several amendments following your comments @AKuederle, and I improved other parts as well. I refactored the main function to class structure. Do you think it is good to refactor the secondary function as well (correct_sensor_orientation_dynamic)?
Regarding testing, we could add the data under 'example_data' and some tests under 'tests' when I have the files ready. I need to do more work on this because the work I have done in creating 8 'wrong' files does not always lead to walking bouts.

DMegaritis committed:
rebased to main

DMegaritis committed:
Updated imports

DMegaritis committed:
Updated imports

DMegaritis committed:
Updated column names to "acc_is", "acc_ml", "acc_pa", "gyr_is", "gyr_ml", "gyr_pa"

DMegaritis committed:
Updated column names to "acc_is", "acc_ml", "acc_pa", "gyr_is", "gyr_ml", "gyr_pa"

DMegaritis committed:
first implementation of the re-orientation example

DMegaritis committed:
first implementation of the re-orientation example

DMegaritis committed:
first implementation of the re-orientation example

DMegaritis committed:
first implementation of the re-orientation example

DMegaritis committed:
first implementation of the re-orientation example

DMegaritis committed:
first implementation of the re-orientation example

DMegaritis committed:
Merge remote-tracking branch 'origin/main' into Reorientation

DMegaritis committed:
GS is called with the whole data rather than with acc alone

DMegaritis committed:
GS is called with the whole data rather than with acc alone

DMegaritis committed:
Implemented examples with disoriented data

DMegaritis committed:
Implemented examples with disoriented data

DMegaritis committed:
Disoriented data

DMegaritis committed:
Disoriented data

DMegaritis commented:
Some comments:
I believe there might be a typo in the file` .venv/lib/python3.11/site-packages/gaitmap/utils/consts.py.` The columns are listed as acc_si and gyr_si, but they should be acc_is and gyr_is.

Additionally, within the Madgwick, why is the function `is_single_sensor_data` called with frame="sensor" instead of frame="body"? Also, shouldn't data[SF_***] be data[BF_***]?

DMegaritis committed:
Merge remote-tracking branch 'origin/Reorientation' into Reorientation

DMegaritis committed:
test

DMegaritis committed:
Merge remote-tracking branch 'origin/main' into Reorientation

DMegaritis committed:
Updated acceleration function to return the data with the same column names as original data

DMegaritis committed:
Updated acceleration function to return the data with the same column names as original data

DMegaritis committed:
Updated to use mobgap madgwick function

DMegaritis committed:
Updated plot syntax

DMegaritis committed:
Updated GSD to Ionescu which is orientation agnostic

DMegaritis committed:
Corrected plots

DMegaritis committed:
Corrected wrong orientation

DMegaritis committed:
Corrected wrong orientation

DMegaritis committed:
Corrected wrong orientation

DMegaritis committed:
Updated comments

DMegaritis committed:
Updated comments

DMegaritis committed:
Updated indexing because a single datapoint at the end of each bout was not included in the slicing this point was not included in the operations

DMegaritis committed:
Updated indexing because a single datapoint at the end of each bout was not included in the slicing this point was not included in the operations

DMegaritis commented:
I replaced the GSD algorithm used in the reorientation block from Iluz to Ionescu, which is orientation-agnostic. This ensures that walking bouts are analysed separately, as originally intended.

DMegaritis committed:
Removed local figure saving

DMegaritis committed:
Marked branch as deprecated and updated README with pointer to a maintained implementation.

DMegaritis committed:
Marked branch as deprecated and updated README with pointer to a maintained implementation.
